File: src/model_landing_3dof.cpp
#include "model_landing_3dof.h"



model_landing_3dof::StateVector model_landing_3dof::ode(const StateVector &x, const ControlVector &u) {

    const double throttle = u(0, 0);
    const double gimbalAngle = u(1, 0);
    const double vx = x(2, 0);
    const double vy = x(3, 0);
    const double theta = x(4, 0);
    const double dtheta = x(5, 0);
    const double x0 = TWR*g*throttle;
    const double x1 = gimbalAngle + theta;

    StateVector f;
    f(0, 0) = vx;
    f(1, 0) = vy;
    f(2, 0) = x0*sin(x1);
    f(3, 0) = g*(TWR*throttle*cos(x1) - 1);
    f(4, 0) = dtheta;
    f(5, 0) = -rTB*x0*sin(gimbalAngle)/(rG*rG);


    return f;
}

model_landing_3dof::StateMatrix model_landing_3dof::state_jacobian(const StateVector &x, const ControlVector &u) {

    const double throttle = u(0, 0);
    const double gimbalAngle = u(1, 0);
    const double theta = x(4, 0);
    const double x0 = TWR*g*throttle;
    const double x1 = gimbalAngle + theta;

    StateMatrix A;
    A.setZero();
    A(0, 2) = 1;
    A(1, 3) = 1;
    A(2, 4) = x0*cos(x1);
    A(3, 4) = -x0*sin(x1);
    A(4, 5) = 1;


    return A;
}



model_landing_3dof::ControlMatrix model_landing_3dof::control_jacobian(const StateVector &x, const ControlVector &u) {

    const double throttle = u(0, 0);
    const double gimbalAngle = u(1, 0);
    const double theta = x(4, 0);
    const double x0 = TWR*g;
    const double x1 = gimbalAngle + theta;
    const double x2 = x0*sin(x1);
    const double x3 = x0*cos(x1);
    const double x4 = TWR*g*rTB/(rG*rG);

    ControlMatrix B;
    B.setZero();
    B(2, 0) = x2;
    B(2, 1) = throttle*x3;
    B(3, 0) = x3;
    B(3, 1) = -throttle*x2;
    B(5, 0) = -x4*sin(gimbalAngle);
    B(5, 1) = -throttle*x4*cos(gimbalAngle);


    return B; 
}



void model_landing_3dof::add_application_constraints(EcosWrapper &solver, size_t K) {
    // TODO
}

model_landing_3dof::StateVector model_landing_3dof::get_random_state() {
    StateVector X;
    X.setRandom();
    X *= 10;
    return X;
}

model_landing_3dof::ControlVector model_landing_3dof::get_random_input() {
    ControlVector U;
    U.setRandom();
    return U;
}

File: src/model_landing_3dof.py
from sympy import *
import sympy


def matrix_c_code(M, type, name, inputs):
    result = ""

    for input_name, input_vector in inputs.items():
        for i, input_sym in enumerate(input_vector):
            if sum([len(e.find(input_sym)) for e in M]) > 0:
                result += "    const double " + str(input_sym) + " = " + input_name + "(" + str(i) + ", 0);\n"

    replacements, M = sympy.cse(M)
    M = Matrix(M)

    for lhs, rhs in replacements:
        result += "    const double " + sympy.ccode(lhs) + " = " + sympy.ccode(rhs) + ";\n"

    result += "\n    " + type + " " + name + ";\n    " + name + ".setZero();\n"

    for i in range(M.rows):
        for j in range(M.cols):
            if(M[i,j] != 0):
                result += "    " + name + "(" + str(i) + ", " + str(j) + ") = "
                result += sympy.ccode(M[i,j]) + ";\n"
    return result

def c_code_postprocessing(code):
    code = code.replace("pow(rG, 2)", "(rG*rG)")
    return code

def main():

    f = sympy.zeros(6,1)

    rx, ry, vx, vy, theta, dtheta = symbols('rx ry vx vy theta dtheta')
    x = Matrix([rx, ry, vx, vy, theta, dtheta])

    throttle, gimbalAngle = symbols('throttle gimbalAngle')
    u = [throttle, gimbalAngle]

    TWR, g, rG, rTB = symbols('TWR g rG rTB')

    f[0,0] = vx
    f[1,0] = vy
    f[2,0] = g * TWR * sympy.sin(theta + gimbalAngle)
    f[3,0] = g * (TWR * sympy.cos(theta + gimbalAngle) - 1)
    f[4,0] = dtheta
    f[5,0] = -g * TWR * sympy.sin(gimbalAngle) * rTB / sympy.Mul(rG, rG, evaluate=False)

    inputs = {'x':x, 'u':u}



    A = f.jacobian(x)
    B = f.jacobian(u)



    print("\node:")
    print(c_code_postprocessing(matrix_c_code(f, "StateVector", "f", inputs)))

    print("\nstate_jacobian:")
    print(c_code_postprocessing(matrix_c_code(A, "StateMatrix", "A", inputs)))

    print("\ncontrol_jacobian:")
    print(c_code_postprocessing(matrix_c_code(B, "ControlMatrix", "B", inputs)))

    


if __name__ == "__main__": main()